### PR TITLE
[FSDP] add no_broadcast_optim_state option

### PR DESCRIPTION
**Pull request by sshleifer**
### Problem
- some parameters are large and their optimizer state cannot be gathered, for example, the experts in mixture of experts.
- We need a way to allow callers to control which keys in the optimizer state dict (`OSD`) are collected to avoid OOM.
  - `rebuild_full_params` already implements this, but the approach is recursive.
- Callers can now write custom logic to save `OSD` state of certain submodules without gathering.

### Approach
- We set `m.no_broadcast_optim_state=True` if a child instance `m` has world_size=1 and it's own `process_group`.
- We remove `no_broadcast_optim_state` parameters from each rank's OSD before collection.
- We add back rank 0's expert parameters to OSD after collection.
- This allows us to produce an `OSD` with the correct number of entries and correct shapes for expert parameters, just like `state_dict`
- We let callers overwrite the state for the expert on rank `i` with the correct contents.

### Determination in `_set_is_root`

```python
m.no_broadcast_optim_state = m.no_broadcast_optim_state or (
(m.world_size == 1) and (m.world_size < self.world_size) and (m.process_group != self.process_group))
```
the tests show that we can recognize MOE experts like this and avoid OOM without any change to callers.

### TODO
- Optimization: Do not copy `osd['param_groups']` from each worker.

### CODE REVIEW

**Comment by min-xu-ai** [comment]
@msbaines Mandeep a failure here on a pipeline test is sporadic?

**Review by min-xu-ai — APPROVED** [review]
This is nice. I think I mostly followed but can't say I checked the logic of every line. Hopefully the test is sufficient to cover the new case this enables and ensure no regression. Some comments that are not blocking. The only really issue is the commented out assertion, seems scary.

File: fairscale/nn/data_parallel/fsdp_optim_utils.py [hunk 1]
```diff
@@ -21,22 +21,22 @@ def flatten_optim_state_dict(sd: Dict) -> Dict:
...
     # Now combine all tensors in each buffer using torch.cat().
-    for consolidated_pid, state in new_state.items():
+    for local_id, state in new_state.items():
         for buffer_name, tensors in state.items():
-            new_state[consolidated_pid][buffer_name] = torch.cat(tensors)
-        new_state[consolidated_pid].update(non_tensor_state)
-    new_sd = {"state": new_state, "param_groups": sd["param_groups"]}
+            new_state[local_id][buffer_name] = torch.cat(tensors)
+        new_state[local_id].update(non_tensor_state)
+    new_sd = {"state": new_state, "param_groups": copy.deepcopy(sd["param_groups"])}
```
[the remark this answers is not in the document]
Comment (sshleifer) [reply fairscale/nn/data_parallel/fsdp_optim_utils.py:39]: Luckily there are no tensors in param_groups, but that's very useful to know!

File: fairscale/nn/data_parallel/fully_sharded_data_parallel.py [hunk 2]
```diff
@@ -157,6 +157,9 @@ class FullyShardedDataParallel(nn.Module):
             device, the param's device will be used. If not given and module
             params are on CPU, then the current CUDA device (as indicated by
             ``torch.cuda.current_device()`` will be used.
+        no_broadcast_optim_state: (bool, Optional)
+            do not broadcast this modules optimizer state when ``gather_full_optim_state_dict`` is called
```
[the remark this answers is not in the document]
Comment (sshleifer) [reply fairscale/nn/data_parallel/fully_sharded_data_parallel.py:161]: Yes, will document!

File: tests/nn/data_parallel/test_fsdp_optimizer_utils.py [hunk 3]
```diff
@@ -36,11 +36,12 @@ def assert_equal(a, b):
 
 class TestOptimizerUtils(DistributedTest):
     @parameterized.expand(
-        [[functools.partial(SGD, momentum=0.9), True], [SGD, False], [Adam, False], [Adadelta, True]],
+        [[functools.partial(SGD, momentum=0.9), True], [SGD, False], [Adam, False], [Adadelta, True], [Adam, True]],
         name_func=rename_test,
     )
     def test_consolidate_optimizer(self, optim_fn, transformer):
         config = {"mixed_precision": True, "flatten_parameters": True}
+        config["compute_dtype"] = torch.float32
```
Comment (myleott) [inline tests/nn/data_parallel/test_fsdp_optimizer_utils.py:44]: I'm guessing this and the autocast change where required to get tests to pass?

File: tests/nn/data_parallel/test_fsdp_optimizer_utils.py [hunk 4]
```diff
@@ -89,6 +90,12 @@ def _test_consolidated_optimizer(self, config, rank, group, optim_fn=torch.optim
 
         if fsdp.rank > 0:
             return
```
Comment (myleott) [inline tests/nn/data_parallel/test_fsdp_optimizer_utils.py:92]: should we also test the non-0 ranks at some point?

File: fairscale/nn/data_parallel/fsdp_optim_utils.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (myleott) [reply fairscale/nn/data_parallel/fsdp_optim_utils.py:39]: +1 to adding a comment about param groups not having tensors (thus deepcopy being okay)

**Review by myleott — APPROVED** [review]
LGTM, although I will need to spend some time stepping through the code at some point to fully understand the implementation 😄 

Should we also test that `no_broadcast_optim_state` is inferred properly for the MixtureOfExperts model?

**Comment by msbaines** [comment]
> @msbaines Mandeep a failure here on a pipeline test is sporadic?

Yep. Here is a fix: https://github.com/facebookresearch/fairscale/pull/575

File: tests/nn/data_parallel/test_fsdp_optimizer_utils.py (hunk 4 above)
Comment (sshleifer) [reply tests/nn/data_parallel/test_fsdp_optimizer_utils.py:92]: added `assert sd is None` before return

File: tests/nn/data_parallel/test_fsdp_optimizer_utils.py (hunk 3 above)
Comment (sshleifer) [reply tests/nn/data_parallel/test_fsdp_optimizer_utils.py:44]: yes, I followed the logic of [`__test_identical_outputs`](https://github.com/facebookresearch/fairscale/blob/master/tests/nn/data_parallel/test_fsdp.py#L81)

File: fairscale/nn/data_parallel/fully_sharded_data_parallel.py [hunk 5]
```diff
@@ -841,6 +850,12 @@ def _set_is_root(self) -> None:
                 if m.process_group != self.process_group:
                     self.children_share_process_group = False
 
+                # if child instance in its own (smaller) world, that was probably an attempt to avoid OOM.
+                # Therefore gathering this child's optim state will probably avoid OOM, so we won't do it.
```
[the remark this answers is not in the document]
Comment (sshleifer) [reply fairscale/nn/data_parallel/fully_sharded_data_parallel.py:854]: yes, nice catch!

File: fairscale/nn/data_parallel/fsdp_optim_utils.py [hunk 6]
```diff
@@ -53,7 +53,8 @@ def check_param_counts_before_sharding(full_optim_state_dict: Dict, n_instances:
         f"there were {n_local_params_in_opt}"
     )
     stateless = len(full_optim_state_dict["state"]) == 0
-    assert stateless or (n_instances == n_local_params_in_opt), msg
+    if not (stateless or (n_instances == n_local_params_in_opt)):
+        print(msg)
```
[the remark this answers is not in the document]
Comment (sshleifer) [reply fairscale/nn/data_parallel/fsdp_optim_utils.py:57]: yes, nice catch. I put back the assert.

**Pull request merged**